src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -69,14 +90,50 @@ class PeerManagerActor(
     peer
   }
 
+  def getPeers(): Future[Peers] = {
+    implicit val timeout = Timeout(2.seconds)
+
+    Future.traverse(peers.values) { peer =>
+      (peer.ref ? PeerActor.GetStatus)
+        .mapTo[PeerActor.StatusResponse]
+        .map(sr => (peer, sr.status))
+    }.map(r => Peers.apply(r.toMap))
+  }
+
+  def tryDiscardPeersToFreeUpLimit(): Future[Unit] = {
+    getPeers() flatMap { peers =>
+      val peersToPotentiallyDisconnect = peers.peers.filter {
+        case (_, Handshaking(numRetries)) if numRetries > 0 => true /* still handshaking and retried at least once */
+        case (_, PeerActor.Status.Disconnected) => true /* already disconnected */
+        case _ => false
+      }
+
+      if (peers.peers.size - peersToPotentiallyDisconnect.size >= peerConfiguration.maxPeers) {
+        Future.failed(new RuntimeException("Too many peers"))
```

Comment:
Yes. It's not really that easy. If there are, for example 2 peers that are `Connecting` and you have 2 more peers on the list that you may try connecting to (for example, as in your first case). What should we do?
If we try to connect to them anyway, and they succeed but also the `Connecting` peers succeed, we end up with more peers than the limit.
The logic here is, if we try to connect to new peer, and there's another that's `Connecting` (without any retries) they're treated "equally", i.e that `Connecting` one was first so it should not be disconnected. If at later point this peer fails to connect (or has multiple reconnect retry attempts) it will be considered "worse" and put in `peersToPotentiallyDisconnect` list to free up space. This check will happen every `bootstrap-nodes-scan-interval`